text/0000-compose.md

```diff
@@ -0,0 +1,100 @@
+* 2019-03-18
+* RFC PR: (leave this empty)
+* Fusion Issue: (leave this empty)
+
+# Summary
+
+This adds support for composing plugins via `app.compose`, and updates the `createToken` API to
+add an optional compose function as the second argument.
+
+# Basic example
+
+```js
+const ComposeExampleToken = createToken('ComposeExampleToken', (items) => {
```

Comment:
I'm trying to wrap my head around the implications of adding this additional construct to tokens.  I may want to let this bounce around for a bit, but my immediate thought is that this breaks with the notion that tokens today are effectively a mechanism to reference objects.  Attaching functionality to them breaks this paradigm and I think it's something we want to spend some time thinking about.

Should the `LoggerToken` care about *how* the value it references was constructed?  Is there a prescribed composition of values that applies to all loggers?

Reply:
This is something I thought about also. We could have the compose function also live in the DI system. For example, you could do something like

```js
app.register(Token.compose, composeFn);
```

However, I think in the vast majority of cases, there should be a single way to compose something based on its type. Unary functions have a standard compose function. Redux reducers have a standard compose function. Etc. It makes sense in my head for the token to know how to compose things, because it knows what the types of those things are.

However, we could support overriding the default compose function with my example above if it ever comes to that.